### PR TITLE
introduce clever config subcommand

### LINKED ISSUE #74: Edit application config with `clever config`
For now: name, description, zero-downtime deployment, sticky sessions, dedicated build instances, cancel on push.

`clever config set name <name>`
`clever config set description <description>`
`clever config set zero-downtime (true|false)`
`clever config set sticky-sessions (true|false)`
`clever config set dedicated-build (false|<flavor>)`
`clever config set cancel-on-push (true|false)`

It could also be done with flags, but that could get messy. Thoughts?

### PR DESCRIPTION
**Pull request by Keruspe**
Fixes #74

### CODE REVIEW

**Comment by Keruspe** [comment]
For now, it's read-only. `clever config` gets "all" the registered configuration, `clever config get foo` gets the configuration with id 'foo'.

At this stage, consider this as a RFC before going further :)

**Comment by Keruspe** [comment]
ATM the set command has two arguments: the config name and the config value.

Subcommands could be generated instead from the CONFIG_KEYS if we want. That could make for better --help

File: src/commands/config.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+const CONFIG_KEYS = [
+  { id: 'name', name: 'name', displayName: 'Name' },
+  { id: 'description', name: 'description', displayName: 'Description' },
+  { id: 'zero-downtime', name: 'homogeneous', displayName: 'Zero-downtime deployment', displayer: zeroDowntimeDisplayer },
+  { id: 'sticky-sessions', name: 'stickySessions', displayName: 'Sticky sessions', displayer: booleanDisplayer },
+  { id: 'cancel-on-push', name: 'cancelOnPush', displayName: 'Cancel current deployment on push', displayer: booleanDisplayer },
+];
+
+function getConfigById (id) {
+  const [config] = CONFIG_KEYS.filter((config) => config.id === id);
```
Comment (hsablonniere) [inline src/commands/config.js]:
```suggestion
  const config = CONFIG_KEYS.find((config) => config.id === id);
```

File: src/commands/config.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  if (config === undefined) {
```
Comment (hsablonniere) [inline src/commands/config.js]:
```suggestion
  if (config == null) {
```

File: src/commands/config.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+function printConfig (app, config) {
+  if (app[config.name] !== undefined) {
+    const displayer = config.displayer || defaultDisplayer;
+    Logger.println(`${config.displayName}: ${displayer(app[config.name])}`);
+  }
+}
+
+function printConfiguration (app, configurationName) {
+  if (configurationName === undefined) {
```
Comment (hsablonniere) [inline src/commands/config.js]:
```suggestion
  if (configurationName == null) {
```

File: src/commands/config.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+function booleanDisplayer (value) {
+  return (value) ? 'enabled' : 'disabled';
+}
+
+function defaultDisplayer (value) {
+  return `${value}`;
+}
+
+function printConfig (app, config) {
+  if (app[config.name] !== undefined) {
```
Comment (hsablonniere) [inline src/commands/config.js]:
```suggestion
  if (app[config.name] != null) {
```

File: src/commands/config.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+function printConfiguration (app, configurationName) {
+  if (configurationName === undefined) {
+    for (const config of CONFIG_KEYS) {
+      printConfig(app, config);
+    }
+  }
+  else {
+    const config = getConfigById(configurationName);
+    if (config !== undefined) {
```
Comment (hsablonniere) [inline src/commands/config.js]:
```suggestion
    if (config != null) {
```

File: src/commands/config.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+function zeroDowntimeDisplayer (homogeneous) {
+  return booleanDisplayer(!homogeneous);
+}
+
+function booleanDisplayer (value) {
+  return (value) ? 'enabled' : 'disabled';
+}
+
+function defaultDisplayer (value) {
+  return `${value}`;
```
Comment (hsablonniere) [inline src/commands/config.js]: Is it written this way to cast it to a string?

File: src/commands/config.js [hunk 7]
```diff
@@ -63,4 +79,18 @@ async function get (params) {
...
 }
 
-module.exports = { get };
+async function set (params) {
+  const [configurationName, configurationValue] = params.args;
+  const { alias } = params.options;
+  const { ownerId, appId } = await AppConfig.getAppDetails({ alias });
+  const config = getConfigById(configurationName);
+
+  if (config !== undefined) {
```
Comment (hsablonniere) [inline src/commands/config.js]:
```suggestion
  if (config != null) {
```

File: src/models/application_configuration.js [hunk 8]
```diff
@@ -51,6 +53,63 @@ function parse (config, value) {
   }
 }
 
+function getUpdateOptions () {
+  return CONFIG_KEYS.flatMap((config) => getConfigOptions(config));
+}
+
+function getConfigOptions (config) {
+  switch (config.kind) {
+    case 'bool': case 'inverted-bool': {
```
Comment (hsablonniere) [inline src/models/application_configuration.js]:
```suggestion
    case 'bool':
    case 'inverted-bool': {
```

File: src/commands/config.js (hunk 6 above)
Comment (Keruspe) [reply src/commands/config.js]: Sort of. At first in some other version of the code it was to have different "symmetrical" branches, and I figured it doesn't hurt, we we can just use identity here in fact

File: src/models/application_configuration.js [hunk 9]
```diff
@@ -47,6 +51,9 @@ function parse (config, value) {
     case 'inverted-bool': {
       return (value === 'false');
     }
+    case 'force-https': {
+      return (value === false) ? 'DISABLED' : 'ENABLED';
```
Comment (hsablonniere) [inline src/models/application_configuration.js:55]:
```suggestion
      return (value === 'false') ? 'DISABLED' : 'ENABLED';
```

Here's what I get in my test:

```
clever config set force-https false
Force redirection of HTTP to HTTPS: enabled
```

**Pull request merged**